lib/xctest.js

```diff
@@ -0,0 +1,241 @@
+            NSUnbufferedIO: 'YES',
+            SQLITE_ENABLE_THREAD_ASSERTIONS: '1',
+            WDA_PRODUCT_BUNDLE_IDENTIFIER: '',
+            XCTestConfigurationFilePath: xctestConfigurationPath,
+            XCODE_DBG_XPC_EXCLUSIONS: 'com.apple.dt.xctestSymbolicator',
+            MJPEG_SERVER_PORT: '',
+            USE_PORT: '',
+            // %p means pid
+            LLVM_PROFILE_FILE: appContainer + '/tmp/%p.profraw',
+        };
+        if (majorVersion >= MAJOR_VERSION_11) {
+            appEnv.DYLD_INSERT_LIBRARIES = '/Developer/usr/lib/libMainThreadChecker.dylib';
+            appEnv.OS_ACTIVITY_DT_MODE = 'YES';
+        }
+        const appArgs = [
+            '-NSTreatUnknownArgumentsAsOpen', 'NO',
+            '-ApplePersistenceIgnoreState', 'YES'
+        ];
+        const appOptions = { StartSuspendedKey: false };
+        if (majorVersion >= MAJOR_VERSION_12) {
+            appOptions.ActivateSuspended = true;
+        }
+        const pid = await instrumentService.callChannel(PROCESS_CONTROL,
+            'launchSuspendedProcessWithDevicePath:bundleIdentifier:environment:arguments:options:',
+            appPath, this.xctestBundleId, appEnv, appArgs, appOptions);
+        log.info(`Pid of launched ${this.xctestBundleId}: ${pid.selector}`);
+        const msg = new DTXMessageAuxBuffer();
+        msg.appendObject(pid.selector);
+        await instrumentService.callChannel(PROCESS_CONTROL, 'startObservingPid:', msg);
+        const logDebug = (message) => {
+            log.debug(`instrument output: ${message.auxiliaries?.[0]}`);
+        };
+        instrumentService.registerSelectorCallback('_XCT_logDebugMessage', logDebug);
+        return { instrumentService, pid: pid.selector };
+    }
+
+    /**
+     * @private
+     * @param {XCTestConfiguration} xctestConfiguration plist contains
+     * @param {string} xctestPath where xctestConfiguration should be place in app sandbox.
+     */
+    async _writeConfigurationToDevice(xctestConfiguration, xctestPath) {
+        const xctestContent = xctestConfiguration.getBytes();
+        const houseArrestService = await startHouseArrestService(this.udid);
+        const vendContainer = await houseArrestService.vendContainer(this.xctestBundleId);
+        const list = await vendContainer.listDirectory(TMP_FOLDER_PREFIX);
+        for (const file of list) {
+            if (file.endsWith(XCTEST_CONFIGURATION_EXTENSION)) {
+                const fullPath = `${TMP_FOLDER_PREFIX}/${file}`;
+                log.debug(`removing ${fullPath}`);
+                await vendContainer.deleteDirectory(fullPath);
+            }
+        }
+        const stream = await vendContainer.createWriteStream(xctestPath, {});
+        try {
+            await new B((resolve, reject) => {
+                stream.write(xctestContent, resolve);
+                stream.on('error', reject);
+            });
+        } finally {
+            stream.end();
```

Comment:
it would probably not work as expected. We need to close each service right after it was opened. For example:

service1.start();
try {
  ...
  service2.start();
  try {
    ...
  } finally {
    service2.stop();
  }
  ....
} finally {
  service1.stop();
}